# Clustering the Neigborhoods of London and Paris

Tiffany Kwan

## 1. Introduction

London and Paris are two of the most popular tourist destinations.  This project will group the two neighborhoods respectively to draw insights into the two cities.

## 2. Business Problem

We want to help tourists choose which neighborhood to visit based on likes and experiences.  Not only will this help tourists, but it will also help individuals who are considering relocating to one of these neigborhoods.  Further, this will help stakeholders makw decisions based on what amenities wach city has to offer.

## 3. Data 

We will use postal codes as a starting point for geographical location data in London and Paris.  The codes will help us find popular venues, neighborhoods and boroughs. 

### London

We will use data from https://en.wikipedia.org/wiki/List_of_areas_of_London

For more information about geographical locations, we will utilize ArcGIS API which will enable us to connect people, locations and data using interactive maps. 

The following will be the variables used

1. borough: Name of Neighborhood
2. town: Name of borough
3. post_code: postal codes for London
4. latitiude: Latitdue for Neighborhood
5. longitude: Longitde for Neighborhood

### Paris

We will use data from JSON data https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e

The following variables will be used

1. postal_code: postal codes for France
2. nom_comm: Name of Neighborhood in France
3. nom_dept: Name of the boroughs
4. feo_point_2d: Tuple containing the latitude and longitude

#### Foursquare API Data

We will use Foursquare data to find data about specific venues in different neighnorhoods.  First, we will find the list of neighborhoods, then connect to Foursquare API Data to find venues inside each neighborhood within a 500 meter radius.  

The following variables will be used

1. Neighbourhood : Name of the Neighbourhood
2. Neighbourhood Latitude : Latitude of the Neighbourhood
3. Neighbourhood Longitude : Longitude of the Neighbourhood
4. Venue : Name of the Venue
5. Venue Latitude : Latitude of Venue
6. Venue Longitude : Longitude of Venue
7. Venue Category : Category of Venue

All of this collected data will help us build our model 

## Methodology

Import all the required packages

In [1]:
!pip install folium

     |████████████████████████████████| 93 kB 4.2 MB/s  eta 0:00:01


In [2]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium

# import k-means for the clustering stage
from sklearn.cluster import KMeans

These packages will help us explore each city individually. Also, we would be able to plot the map to show the separate neighborhoods, then build a model based on similar characteristics.  

## Exploring London

Start by obtaining the list of areas of london

In [3]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

We were able to make the connection

In [4]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
             Location                     London borough       Post town  \
 0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
 1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
 2          Addington                         Croydon[8]         CROYDON   
 3         Addiscombe                         Croydon[8]         CROYDON   
 4        Albany Park                             Bexley  BEXLEY, SIDCUP   
 ..               ...                                ...             ...   
 527         Woolwich                          Greenwich          LONDON   
 528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
 529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
 530          Yeading                         Hillingdon           HAYES   
 

We need to select the second table

In [5]:
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
527         Woolwich                          Greenwich          LONDON   
528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
530          Yeading                         Hillingdon           HAYES   
531         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
527              SE18       020    TQ435795  
528               KT4       020    TQ225655  
529               W12       020    TQ225815  
530               UB4       020    TQ115825  
531               UB7       020    TQ063804  

[532 rows x 6 columns]

## Data Preprocessing

Remove the spaces and add _ in between words

In [6]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location                     London borough       Post_town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
527         Woolwich                          Greenwich          LONDON   
528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
530          Yeading                         Hillingdon           HAYES   
531         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS_grid_ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
527              SE18       020    TQ435795  
528               KT4       020    TQ225655  
529               W12       020    TQ225815  
530               UB4       020    TQ115825  
531               UB7       020    TQ063804  

[532 rows x 6 columns]

Some columns contain special characters so there is not _ between the words.

## Feature Selection

We can drop locations, dial codes and OS grid

In [7]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

In [8]:
df1.head()


London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

Rename the Postcode district column

In [9]:

df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
527                          Greenwich          LONDON       SE18
528       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
529             Hammersmith and Fulham          LONDON        W12
530                         Hillingdon           HAYES        UB4
531                         Hillingdon    WEST DRAYTON        UB7

[532 rows x 3 columns]

Remove the brackets

In [10]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
527                       Greenwich          LONDON       SE18
528    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
529          Hammersmith and Fulham          LONDON        W12
530                      Hillingdon           HAYES        UB4
531                      Hillingdon    WEST DRAYTON        UB7

[532 rows x 3 columns]

Dimensions of the dataframe:

In [11]:
df1.shape


(532, 3)

We have 533 records and 3 columns of data

## Feature Engineering

Currently, we want to focus on the neighborhoods of London

In [12]:

df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
522                       Redbridge                  LONDON  IG8, E18
523       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
526                          Barnet                  LONDON       N12
527                       Greenwich                  LONDON      SE18
529          Hammersmith and Fulham                  LONDON       W12

[309 rows x 3 columns]

In [13]:
df1.shape

(309, 3)

Now with 309 rows, we will try to get some descriptive statistics

In [14]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 309 entries, 0 to 529
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    309 non-null    object
 1   town       309 non-null    object
 2   post_code  309 non-null    object
dtypes: object(3)
memory usage: 9.7+ KB


## Geolocations of the London Neighborhoods

### ArcGis API

Find the geographical coordinates for the enighborhoods in order to plot out the map

In [15]:
!pip install arcgis

In [16]:

from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/arcgis/features/_data/geodataset/utils.py:16: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  pd.datetime,


Define London arcgis to return latitude and longitude

In [17]:

def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

In [18]:
c = get_x_y_uk('SE2')

In [19]:
c

'51.492450000000076,0.12127000000003818'

Copy over the postal codes and pus it to the geolocator function

In [20]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
522    IG8, E18
523         IG8
526         N12
527        SE18
529         W12
Name: post_code, Length: 309, dtype: object

Pass the postal codes to get geographical coordinates

In [21]:

coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
522    51.589770000000044,0.030520000000024083
523      51.50642000000005,-0.1272099999999341
526     51.615920000000074,-0.1767399999999384
527      51.48207000000008,0.07143000000002075
529      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 309, dtype: object

### Latitude

Extract the Latitude

In [22]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
522    51.589770000000044
523     51.50642000000005
526    51.615920000000074
527     51.48207000000008
529     51.50645000000003
Name: post_code, Length: 309, dtype: object

### Longitude

Extract the Longitude

In [23]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
522    0.030520000000024083
523     -0.1272099999999341
526     -0.1767399999999384
527     0.07143000000002075
529     -0.2369099999999662
Name: post_code, Length: 309, dtype: object

Now that we have the geographical coordinates of London's neighborhoods, we can merger our source data with the coordinates

In [24]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged


borough                    town post_code  \
0                Bexley, Greenwich                   LONDON       SE2   
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4   
6                              City                  LONDON       EC3   
7                       Westminster                  LONDON       WC2   
9                           Bromley                  LONDON      SE20   
..                              ...                     ...       ...   
522                       Redbridge                  LONDON  IG8, E18   
523       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8   
526                          Barnet                  LONDON       N12   
527                       Greenwich                  LONDON      SE18   
529          Hammersmith and Fulham                  LONDON       W12   

     latitude  longitude  
0    51.49245    0.12127  
1    51.51324   -0.26746  
6    51.51200   -0.08058  
7    51.51651   -0.11968  
9    51.41009   -0.05683  
..        ...        ...  
522  51.58977    0.03052  
523  51.50642   -0.12721  
526  51.61592   -0.17674  
527  51.48207    0.07143  
529  51.50645   -0.23691  

[309 rows x 5 columns]

In [25]:
london_merged.dtypes


borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

### Coordinates of London

In [26]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [27]:
london_lat_coords


51.50642000000005

This is london's geocode

### Visualize the Map of London

Use the folium package to visualize the Map of London

In [28]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

### Venues in London

We will now use FOursquare API to get the venue and venue categories around each neighborhood in London

In [29]:
CLIENT_ID = 'BLN5212LPUOTLJ5ZQGYTKZBR2PQYQTOAKSAS4R15KALGEE0D' 
CLIENT_SECRET = 'PPJXW2LHVZ0HFJGLTTOFZ43KYMVDAC3WYMEYRLN4PY1O2WR4'
VERSION = '20180605' # Foursquare API version

Now, obtain venue categories

In [30]:
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Get venues in London

In [31]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])



Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and ChelseaHammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames
En

Now, it is time to sample data

In [32]:
venues_in_London.head()


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue Venue Category  
0                      Lesnes Abbey  Historic Site  
1                       Sainsbury's    Supermarket  
2                              Lidl    Supermarket  
3  Abbey Wood Railway Station (ABW)  Train Station  
4                       Bean @ Work    Coffee Shop

In [33]:
venues_in_London.shape


(10882, 5)

### Grouping by Venue Categories

In [34]:
venues_in_London.groupby('Venue Category').max()


Neighbourhood  \
Venue Category                                                      
Adult Boutique                                          Islington   
African Restaurant                                    Westminster   
American Restaurant                                Waltham Forest   
Antique Shop                                          Westminster   
Arcade                                                Westminster   
...                                                           ...   
Wings Joint                                Hammersmith and Fulham   
Women's Store        Kensington and ChelseaHammersmith and Fulham   
Xinjiang Restaurant                                     Southwark   
Yoga Studio                                           Westminster   
Zoo Exhibit                                                Camden   

                     Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                         
Adult Boutique                     51.52969                 -0.08697   
African Restaurant                 51.52587                 -0.00754   
American Restaurant                51.63261                  0.02795   
Antique Shop                       51.51651                 -0.11968   
Arcade                             51.51651                 -0.11968   
...                                     ...                      ...   
Wings Joint                        51.54187                 -0.19795   
Women's Store                      51.55457                 -0.11478   
Xinjiang Restaurant                51.47480                 -0.09313   
Yoga Studio                        51.55457                 -0.03558   
Zoo Exhibit                        51.53354                 -0.14606   

                                           Venue  
Venue Category                                    
Adult Boutique       Sh! Women's Erotic Emporium  
African Restaurant            Red Sea Restaurant  
American Restaurant                  Spielburger  
Antique Shop            The London Silver Vaults  
Arcade                        Novelty Automation  
...                                          ...  
Wings Joint                             Wingmans  
Women's Store                 Vivien of Holloway  
Xinjiang Restaurant                    Silk Road  
Yoga Studio                            yogahaven  
Zoo Exhibit                        Penguin Beach  

[299 rows x 4 columns]

This area is very diverse: 299 records

### One Hot Encoding

Time to encode venue categories in order to cluster

In [35]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat


Adult Boutique  African Restaurant  American Restaurant  Antique Shop  \
0                   0                   0                    0             0   
1                   0                   0                    0             0   
2                   0                   0                    0             0   
3                   0                   0                    0             0   
4                   0                   0                    0             0   
...               ...                 ...                  ...           ...   
10877               0                   0                    0             0   
10878               0                   0                    0             0   
10879               0                   0                    0             0   
10880               0                   0                    0             0   
10881               0                   0                    0             0   

       Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0           0                 0                       0            0   
1           0                 0                       0            0   
2           0                 0                       0            0   
3           0                 0                       0            0   
4           0                 0                       0            0   
...       ...               ...                     ...          ...   
10877       0                 0                       0            0   
10878       0                 0                       0            0   
10879       0                 0                       0            0   
10880       0                 0                       0            0   
10881       0                 0                       0            0   

       Art Museum  Arts & Crafts Store  ...  Warehouse Store  Water Park  \
0               0                    0  ...                0           0   
1               0                    0  ...                0           0   
2               0                    0  ...                0           0   
3               0                    0  ...                0           0   
4               0                    0  ...                0           0   
...           ...                  ...  ...              ...         ...   
10877           0                    0  ...                0           0   
10878           0                    0  ...                0           0   
10879           0                    0  ...                0           0   
10880           0                    0  ...                0           0   
10881           0                    0  ...                0           0   

       Whisky Bar  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0               0         0          0            0              0   
1               0         0          0            0              0   
2               0         0          0            0              0   
3               0         0          0            0              0   
4               0         0          0            0              0   
...           ...       ...        ...          ...            ...   
10877           0         0          0            0              0   
10878           0         0          0            0              0   
10879           0         0          0            0              0   
10880           0         0          0            0              0   
10881           0         0          0            0              0   

       Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0                        0            0            0  
1                        0            0            0  
2                        0            0            0  
3                        0            0            0  
4                        0            0            0  
...                    ...          ...          ...  
10877                    0            0   

In [36]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  Adult Boutique  African Restaurant  \
0  Bexley, Greenwich                0                   0   
1  Bexley, Greenwich                0                   0   
2  Bexley, Greenwich                0                   0   
3  Bexley, Greenwich                0                   0   
4  Bexley, Greenwich                0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                    0             0       0                 0   
1                    0             0       0                 0   
2                    0             0       0                 0   
3                    0             0       0                 0   
4                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  Warehouse Store  \
0                       0            0           0  ...                0   
1                       0            0           0  ...                0   
2                       0            0           0  ...                0   
3                       0            0           0  ...                0   
4                       0            0           0  ...                0   

   Water Park  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0           0           0         0          0            0              0   
1           0           0         0          0            0              0   
2           0           0         0          0            0              0   
3           0           0         0          0            0              0   
4           0           0         0          0            0              0   

   Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0                    0            0            0  
1                    0            0            0  
2                    0            0            0  
3                    0            0            0  
4                    0            0            0  

[5 rows x 300 columns]

### Venue categories mean value

In [37]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()


Neighbourhood  Adult Boutique  African Restaurant  \
0                 Barnet             0.0                 0.0   
1  Barnet, Brent, Camden             0.0                 0.0   
2                 Bexley             0.0                 0.0   
3      Bexley, Greenwich             0.0                 0.0   
4     Bexley, Greenwich              0.0                 0.0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0             0.006536           0.0     0.0               0.0   
1             0.000000           0.0     0.0               0.0   
2             0.000000           0.0     0.0               0.0   
3             0.000000           0.0     0.0               0.0   
4             0.000000           0.0     0.0               0.0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  Warehouse Store  \
0                0.006536          0.0         0.0  ...              0.0   
1                0.000000          0.0         0.0  ...              0.0   
2                0.000000          0.0         0.0  ...              0.0   
3                0.000000          0.0         0.0  ...              0.0   
4                0.000000          0.0         0.0  ...              0.0   

   Water Park  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0    0.006536         0.0       0.0        0.0          0.0            0.0   
1    0.000000         0.0       0.0        0.0          0.0            0.0   
2    0.000000         0.0       0.0        0.0          0.0            0.0   
3    0.000000         0.0       0.0        0.0          0.0            0.0   
4    0.000000         0.0       0.0        0.0          0.0            0.0   

   Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0                  0.0          0.0          0.0  
1                  0.0          0.0          0.0  
2                  0.0          0.0          0.0  
3                  0.0          0.0          0.0  
4                  0.0          0.0          0.0  

[5 rows x 300 columns]

In [38]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [39]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

### Top Venue Categories

In [40]:

# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden           Music Venue           Bus Station   
2                 Bexley           Supermarket                  Park   
3      Bexley, Greenwich           Sports Club     Convenience Store   
4     Bexley, Greenwich            Supermarket                   Pub   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Grocery Store                   Pub    Italian Restaurant   
1        Clothing Store  Gym / Fitness Center           Supermarket   
2         Historic Site                   Pub         Train Station   
3           Golf Course                  Park         Historic Site   
4                  Park         Train Station           Coffee Shop   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0          Turkish Restaurant              Bus Stop           Supermarket   
1               Historic Site         Fishing Store  Fast Food Restaurant   
2                 Coffee Shop     Convenience Store              Bus Stop   
3  Construction & Landscaping              Bus Stop           Flower Shop   
4           Convenience Store         Historic Site           Film Studio   

        9th Most Common Venue 10th Most Common Venue  
0                    Pharmacy       Sushi Restaurant  
1         Filipino Restaurant            Film Studio  
2  Construction & Landscaping            Golf Course  
3                 Flea Market          Fishing Store  
4          Falafel Restaurant         Farmers Market

### Model Building

Use K Means clustering to cluster the city of London to 5

In [41]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(n_clusters=5, random_state=0)

In [42]:
kmeans_london.labels_


array([0, 4, 3, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0], dtype=int32)

In [43]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_ +1)

In [44]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough    town post_code  latitude  longitude  \
0              Bexley, Greenwich   LONDON       SE2  51.49245    0.12127   
1  Ealing, Hammersmith and Fulham  LONDON    W3, W4  51.51324   -0.26746   
6                            City  LONDON       EC3  51.51200   -0.08058   
7                     Westminster  LONDON       WC2  51.51651   -0.11968   
9                         Bromley  LONDON      SE20  51.41009   -0.05683   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               4           Supermarket                   Pub   
1               2         Grocery Store     Indian Restaurant   
6               1                 Hotel           Coffee Shop   
7               1                 Hotel           Coffee Shop   
9               1           Supermarket     Fish & Chips Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park         Train Station           Coffee Shop   
1         Train Station        Breakfast Spot                  Park   
6    Italian Restaurant  Gym / Fitness Center                   Pub   
7                  Café        Sandwich Place                   Pub   
9                 Hotel  Fast Food Restaurant         Grocery Store   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0     Convenience Store         Historic Site            Film Studio   
1           Zoo Exhibit     Fish & Chips Shop     Falafel Restaurant   
6            Restaurant        Sandwich Place               Wine Bar   
7    Italian Restaurant               Theater             Restaurant   
9     Convenience Store                  Park  Portuguese Restaurant   

        9th Most Common Venue 10th Most Common Venue  
0          Falafel Restaurant         Farmers Market  
1              Farmers Market   Fast Food Restaurant  
6                Cocktail Bar     Falafel Restaurant  
7           French Restaurant           Burger Joint  
9  Construction & Landscaping   Caribbean Restaurant

Drop the NaN values to prevent data skew

In [45]:
london_data_nonan = london_data.dropna(subset=['Cluster Labels'])


### Visualizing clustered neighborhood

In [46]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

### Examining our clusters

Cluster 1

In [47]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]


town  Cluster Labels 1st Most Common Venue  \
6                    LONDON               1                 Hotel   
7                    LONDON               1                 Hotel   
9                    LONDON               1           Supermarket   
10                   LONDON               1           Coffee Shop   
12                   LONDON               1           Coffee Shop   
..                      ...             ...                   ...   
522                  LONDON               1                  Café   
523  LONDON, WOODFORD GREEN               1                  Café   
526                  LONDON               1           Coffee Shop   
527                  LONDON               1                   Pub   
529                  LONDON               1                   Pub   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6             Coffee Shop    Italian Restaurant  Gym / Fitness Center   
7             Coffee Shop                  Café        Sandwich Place   
9       Fish & Chips Shop                 Hotel  Fast Food Restaurant   
10                    Pub                  Café            Food Truck   
12                    Pub                  Café            Food Truck   
..                    ...                   ...                   ...   
522                   Pub         Grocery Store           Coffee Shop   
523                 Hotel                   Pub               Theater   
526                  Café         Grocery Store                   Pub   
527         Grocery Store              Bus Stop           Coffee Shop   
529                  Café           Coffee Shop         Grocery Store   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6                      Pub            Restaurant        Sandwich Place   
7                      Pub    Italian Restaurant               Theater   
9            Grocery Store     Convenience Store                  Park   
10   Vietnamese Restaurant  Gym / Fitness Center    Italian Restaurant   
12   Vietnamese Restaurant  Gym / Fitness Center    Italian Restaurant   
..                     ...                   ...                   ...   
522                 Bakery         Metro Station              Pharmacy   
523    Monument / Landmark                 Plaza                Garden   
526     Italian Restaurant    Turkish Restaurant              Bus Stop   
527      Indian Restaurant    Turkish Restaurant                  Café   
529              Gastropub                 Hotel           Pizza Place   

     8th Most Common Venue       9th Most Common Venue  \
6                 Wine Bar                Cocktail Bar   
7               Restaurant           French Restaurant   
9    Portuguese Restaurant  Construction & Landscaping   
10                    Park                Cocktail Bar   
12                    Park                Cocktail Bar   
..                     ...                         ...   
522              BBQ Joint                        Park   
523      French Restaurant                  Steakhouse   
526            Supermarket                    Pharmacy   
527      Fish & Chips Shop              Clothing Store   
529                   Park             Thai Restaurant   

         10th Most Common Venue  
6            Falafel Restaurant  
7                  Burger Joint  
9          Caribbean Restaurant  
10               Breakfast Spot  
12               Breakfast Spot  
..                          ...  
522          Seafood Restaurant  
523                 Art Gallery  
526            Sushi Restaurant  
527  Construction & Landscaping  
529                      Bakery  

[301 rows x 12 columns]

Cluuster 2

In [48]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 2, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]


town  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  LONDON               2         Grocery Store     Indian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1         Train Station        Breakfast Spot                  Park   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1           Zoo Exhibit     Fish & Chips Shop    Falafel Restaurant   

  9th Most Common Venue 10th Most Common Venue  
1        Farmers Market   Fast Food Restaurant

Cluster 3

In [49]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]


town  Cluster Labels 1st Most Common Venue  \
378  HARROW, STANMOREEDGWARE, LONDON               3                   Gym   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
378                Bakery         Metro Station    Falafel Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
378        Farmers Market  Fast Food Restaurant   Filipino Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
378           Film Studio     Fish & Chips Shop            Fish Market

Cluster 4

In [50]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 4, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]


town  Cluster Labels 1st Most Common Venue  \
0                 LONDON               4           Supermarket   
45   BEXLEYHEATH, LONDON               4           Supermarket   
124               LONDON               4           Supermarket   
291       LONDON, SIDCUP               4           Supermarket   
506               LONDON               4           Supermarket   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                     Pub                  Park         Train Station   
45                   Park         Historic Site                   Pub   
124                  Park         Historic Site                   Pub   
291                  Park         Historic Site                   Pub   
506                  Park         Historic Site                   Pub   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Coffee Shop     Convenience Store         Historic Site   
45          Train Station           Coffee Shop     Convenience Store   
124         Train Station           Coffee Shop     Convenience Store   
291         Train Station           Coffee Shop     Convenience Store   
506         Train Station           Coffee Shop     Convenience Store   

    8th Most Common Venue       9th Most Common Venue 10th Most Common Venue  
0             Film Studio          Falafel Restaurant         Farmers Market  
45               Bus Stop  Construction & Landscaping            Golf Course  
124              Bus Stop  Construction & Landscaping            Golf Course  
291              Bus Stop  Construction & Landscaping            Golf Course  
506              Bus Stop  Construction & Landscaping            Golf Course

Cluster 5

In [51]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 5, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]


town  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
121  LONDON               5           Music Venue           Bus Station   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
121        Clothing Store  Gym / Fitness Center           Supermarket   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
121         Historic Site         Fishing Store  Fast Food Restaurant   

    9th Most Common Venue 10th Most Common Venue  
121   Filipino Restaurant            Film Studio

## Exploring Paris

### Neighborhoods of Paris

### Data Collection

In [52]:
!wget -q -O 'france-data.json' https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e
print("Data Downloaded!")

Data Downloaded!


In [53]:
paris_raw = pd.read_json('france-data.json')
paris_raw.head()

datasetid  \
0  correspondances-code-insee-code-postal   
1  correspondances-code-insee-code-postal   
2  correspondances-code-insee-code-postal   
3  correspondances-code-insee-code-postal   
4  correspondances-code-insee-code-postal   

                                   recordid  \
0  2bf36b38314b6c39dfbcd09225f97fa532b1fc45   
1  7ee82e74e059b443df18bb79fc5a19b1f05e5a88   
2  e2cd3186f07286705ed482a10b6aebd9de633c81   
3  868bf03527a1d0a9defe5cf4e6fa0a730d725699   
4  21e809b1d4480333c8b6fe7addd8f3b06f343e2c   

                                              fields  \
0  {'code_comm': '645', 'nom_dept': 'ESSONNE', 's...   
1  {'code_comm': '133', 'nom_dept': 'SEINE-ET-MAR...   
2  {'code_comm': '378', 'nom_dept': 'ESSONNE', 's...   
3  {'code_comm': '243', 'nom_dept': 'SEINE-ET-MAR...   
4  {'code_comm': '003', 'nom_dept': 'VAL-DE-MARNE...   

                                            geometry  \
0  {'type': 'Point', 'coordinates': [2.2517129721...   
1  {'type': 'Point', 'coordinates': [3.0529405055...   
2  {'type': 'Point', 'coordinates': [2.1971816504...   
3  {'type': 'Point', 'coordinates': [2.7097808131...   
4  {'type': 'Point', 'coordinates': [2.3335102498...   

                record_timestamp  
0  2016-09-21T00:29:06.175+02:00  
1  2016-09-21T00:29:06.175+02:00  
2  2016-09-21T00:29:06.175+02:00  
3  2016-09-21T00:29:06.175+02:00  
4  2016-09-21T00:29:06.175+02:00

### Data Preprocessing

In [54]:
paris_field_data = pd.DataFrame()
for f in paris_raw.fields:
    dict_new = f
    paris_field_data = paris_field_data.append(dict_new, ignore_index=True)

paris_field_data.head()

code_arr code_cant code_comm code_dept code_reg  \
0        3        03       645        91       11   
1        3        20       133        77       11   
2        1        09       378        91       11   
3        5        14       243        77       11   
4        3        34       003        94       11   

                              geo_point_2d  \
0  [48.750443119964764, 2.251712972144151]   
1   [48.41256065214989, 3.052940505560729]   
2    [48.52726809075556, 2.19718165044305]   
3  [48.87307018579678, 2.7097808131278462]   
4   [48.80588035965699, 2.333510249842654]   

                                           geo_shape id_geofla insee_com  \
0  {'type': 'Polygon', 'coordinates': [[[2.238024...     16275     91645   
1  {'type': 'Polygon', 'coordinates': [[[3.076046...     31428     77133   
2  {'type': 'Polygon', 'coordinates': [[[2.203466...     30975     91378   
3  {'type': 'Polygon', 'coordinates': [[[2.727542...     17000     77243   
4  {'type': 'Polygon', 'coordinates': [[[2.343851...     32123     94003   

               nom_comm        nom_dept     nom_region  population  \
0  VERRIERES-LE-BUISSON         ESSONNE  ILE-DE-FRANCE        15.5   
1  COURCELLES-EN-BASSEE  SEINE-ET-MARNE  ILE-DE-FRANCE         0.2   
2             MAUCHAMPS         ESSONNE  ILE-DE-FRANCE         0.3   
3       LAGNY-SUR-MARNE  SEINE-ET-MARNE  ILE-DE-FRANCE        20.2   
4               ARCUEIL    VAL-DE-MARNE  ILE-DE-FRANCE        19.5   

  postal_code            statut  superficie  z_moyen  
0       91370    Commune simple       999.0    121.0  
1       77126    Commune simple      1082.0     88.0  
2       91730    Commune simple       313.0    150.0  
3       77400  Chef-lieu canton       579.0     71.0  
4       94110  Chef-lieu canton       232.0     70.0

### Feature Selection

In [55]:
df_2 = paris_field_data[['postal_code','nom_comm','nom_dept','geo_point_2d']]
df_2

postal_code              nom_comm           nom_dept  \
0          91370  VERRIERES-LE-BUISSON            ESSONNE   
1          77126  COURCELLES-EN-BASSEE     SEINE-ET-MARNE   
2          91730             MAUCHAMPS            ESSONNE   
3          77400       LAGNY-SUR-MARNE     SEINE-ET-MARNE   
4          94110               ARCUEIL       VAL-DE-MARNE   
...          ...                   ...                ...   
1295       77520     CESSOY-EN-MONTOIS     SEINE-ET-MARNE   
1296       93420            VILLEPINTE  SEINE-SAINT-DENIS   
1297       77130         CANNES-ECLUSE     SEINE-ET-MARNE   
1298       78930              VILLETTE           YVELINES   
1299       95270  LE PLESSIS-LUZARCHES         VAL-D'OISE   

                                 geo_point_2d  
0     [48.750443119964764, 2.251712972144151]  
1      [48.41256065214989, 3.052940505560729]  
2       [48.52726809075556, 2.19718165044305]  
3     [48.87307018579678, 2.7097808131278462]  
4      [48.80588035965699, 2.333510249842654]  
...                                       ...  
1295   [48.50730730461658, 3.138844194183689]  
1296   [48.95902025378707, 2.536306342059409]  
1297   [48.36403767307805, 2.990786679832767]  
1298  [48.92627887061508, 1.6937417245662671]  
1299  [49.09572967201378, 2.4547564431234923]  

[1300 rows x 4 columns]

### Feature Engineering

In [56]:
df_paris = df_2[df_2['nom_dept'].str.contains('PARIS')].reset_index(drop=True)
df_paris

postal_code                  nom_comm nom_dept  \
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS   
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS   
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS   
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS   
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS   
5        75019  PARIS-19E-ARRONDISSEMENT    PARIS   
6        75020  PARIS-20E-ARRONDISSEMENT    PARIS   
7        75015  PARIS-15E-ARRONDISSEMENT    PARIS   
8        75005   PARIS-5E-ARRONDISSEMENT    PARIS   
9        75007   PARIS-7E-ARRONDISSEMENT    PARIS   
10       75008   PARIS-8E-ARRONDISSEMENT    PARIS   
11       75013  PARIS-13E-ARRONDISSEMENT    PARIS   
12       75012  PARIS-12E-ARRONDISSEMENT    PARIS   
13       75018  PARIS-18E-ARRONDISSEMENT    PARIS   
14       75001  PARIS-1ER-ARRONDISSEMENT    PARIS   
15       75010  PARIS-10E-ARRONDISSEMENT    PARIS   
16       75016  PARIS-16E-ARRONDISSEMENT    PARIS   
17       75004   PARIS-4E-ARRONDISSEMENT    PARIS   
18       75017  PARIS-17E-ARRONDISSEMENT    PARIS   
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS   

                               geo_point_2d  
0    [48.87689616237872, 2.337460241388529]  
1    [48.86790337886785, 2.344107166658533]  
2    [48.85941549762748, 2.378741060237548]  
3    [48.86305413181178, 2.359361058970589]  
4    [48.84896809191946, 2.332670898588416]  
5    [48.88686862295828, 2.384694327870042]  
6    [48.86318677744551, 2.400819826729021]  
7    [48.84015541860987, 2.293559372435076]  
8   [48.844508659617546, 2.349859385560182]  
9    [48.85608259819694, 2.312438687733857]  
10   [48.87252726662346, 2.312582560420059]  
11   [48.82871768452136, 2.362468228516128]  
12   [48.83515623066034, 2.419807034965275]  
13  [48.892735074561706, 2.348711933867703]  
14    [48.8626304851685, 2.336293446550539]  
15   [48.87602855694339, 2.361112904561707]  
16   [48.86039876035177, 2.262099559395783]  
17  [48.854228281954754, 2.357361938142205]  
18   [48.88733716648682, 2.307485559493426]  
19   [48.82899321160942, 2.327100883257538]

In [57]:
df_paris.shape


(20, 4)

We broke down Paris data to 20 recors, much smaller than London

In [58]:
df_paris.dtypes


postal_code     object
nom_comm        object
nom_dept        object
geo_point_2d    object
dtype: object

### Geolocations of the Neighborhoods of Paris

In [59]:
df_paris['geo_point_2d'][0]


[48.87689616237872, 2.337460241388529]

In [60]:

temp1 = df_paris['geo_point_2d']
temp1

0      [48.87689616237872, 2.337460241388529]
1      [48.86790337886785, 2.344107166658533]
2      [48.85941549762748, 2.378741060237548]
3      [48.86305413181178, 2.359361058970589]
4      [48.84896809191946, 2.332670898588416]
5      [48.88686862295828, 2.384694327870042]
6      [48.86318677744551, 2.400819826729021]
7      [48.84015541860987, 2.293559372435076]
8     [48.844508659617546, 2.349859385560182]
9      [48.85608259819694, 2.312438687733857]
10     [48.87252726662346, 2.312582560420059]
11     [48.82871768452136, 2.362468228516128]
12     [48.83515623066034, 2.419807034965275]
13    [48.892735074561706, 2.348711933867703]
14      [48.8626304851685, 2.336293446550539]
15     [48.87602855694339, 2.361112904561707]
16     [48.86039876035177, 2.262099559395783]
17    [48.854228281954754, 2.357361938142205]
18     [48.88733716648682, 2.307485559493426]
19     [48.82899321160942, 2.327100883257538]
Name: geo_point_2d, dtype: object

In [61]:
paris_latlng = df_paris['geo_point_2d'].astype('str')


### Latitude

In [62]:
paris_lat = paris_latlng.apply(lambda x: x.split(',')[0])
paris_lat = paris_lat.apply(lambda x: x.lstrip('['))
paris_lat

0      48.87689616237872
1      48.86790337886785
2      48.85941549762748
3      48.86305413181178
4      48.84896809191946
5      48.88686862295828
6      48.86318677744551
7      48.84015541860987
8     48.844508659617546
9      48.85608259819694
10     48.87252726662346
11     48.82871768452136
12     48.83515623066034
13    48.892735074561706
14      48.8626304851685
15     48.87602855694339
16     48.86039876035177
17    48.854228281954754
18     48.88733716648682
19     48.82899321160942
Name: geo_point_2d, dtype: object

### Longitude

In [63]:
paris_lng = paris_latlng.apply(lambda x: x.split(',')[1])
paris_lng = paris_lng.apply(lambda x: x.rstrip(']'))
paris_lng

0      2.337460241388529
1      2.344107166658533
2      2.378741060237548
3      2.359361058970589
4      2.332670898588416
5      2.384694327870042
6      2.400819826729021
7      2.293559372435076
8      2.349859385560182
9      2.312438687733857
10     2.312582560420059
11     2.362468228516128
12     2.419807034965275
13     2.348711933867703
14     2.336293446550539
15     2.361112904561707
16     2.262099559395783
17     2.357361938142205
18     2.307485559493426
19     2.327100883257538
Name: geo_point_2d, dtype: object

In [64]:

paris_geo_lat  = pd.DataFrame(paris_lat.astype(float))
paris_geo_lat.columns=['Latitude']
paris_geo_lng = pd.DataFrame(paris_lng.astype(float))
paris_geo_lng.columns=['Longitude']

Preparing our combined data by dropping the geo_point_2d column from our previously stored df_paris and concatenating with the latitude and longitude extracted from it



In [65]:
paris_combined_data = pd.concat([df_paris.drop('geo_point_2d', axis=1), paris_geo_lat, paris_geo_lng], axis=1)
paris_combined_data

postal_code                  nom_comm nom_dept   Latitude  Longitude
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671
5        75019  PARIS-19E-ARRONDISSEMENT    PARIS  48.886869   2.384694
6        75020  PARIS-20E-ARRONDISSEMENT    PARIS  48.863187   2.400820
7        75015  PARIS-15E-ARRONDISSEMENT    PARIS  48.840155   2.293559
8        75005   PARIS-5E-ARRONDISSEMENT    PARIS  48.844509   2.349859
9        75007   PARIS-7E-ARRONDISSEMENT    PARIS  48.856083   2.312439
10       75008   PARIS-8E-ARRONDISSEMENT    PARIS  48.872527   2.312583
11       75013  PARIS-13E-ARRONDISSEMENT    PARIS  48.828718   2.362468
12       75012  PARIS-12E-ARRONDISSEMENT    PARIS  48.835156   2.419807
13       75018  PARIS-18E-ARRONDISSEMENT    PARIS  48.892735   2.348712
14       75001  PARIS-1ER-ARRONDISSEMENT    PARIS  48.862630   2.336293
15       75010  PARIS-10E-ARRONDISSEMENT    PARIS  48.876029   2.361113
16       75016  PARIS-16E-ARRONDISSEMENT    PARIS  48.860399   2.262100
17       75004   PARIS-4E-ARRONDISSEMENT    PARIS  48.854228   2.357362
18       75017  PARIS-17E-ARRONDISSEMENT    PARIS  48.887337   2.307486
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS  48.828993   2.327101

### Coordinates for Paris

In [66]:
paris = geocode(address='Paris, France, FR')[0]
paris_lng_coords = paris['location']['x']
paris_lat_coords = paris['location']['y']
print("The geolocation of Paris: ", paris_lat_coords, paris_lng_coords)

The geolocation of Paris:  48.85341000000005 2.3488000000000397


### Visualize the Map of Paris

In [67]:
# Creating the map of Paris
map_Paris= folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)
map_Paris

# adding markers to map
for latitude, longitude, borough, town in zip(paris_combined_data['Latitude'], paris_combined_data['Longitude'], paris_combined_data['nom_comm'], paris_combined_data['nom_dept']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='Blue',
        fill=True,
        fill_opacity=0.8
        ).add_to(map_Paris)  
    
map_Paris

### Venues in Paris

Get nearby venues present in each neighborhood

In [69]:
venues_in_Paris = getNearbyVenues(paris_combined_data['nom_comm'], paris_combined_data['Latitude'], paris_combined_data['Longitude'])


PARIS-9E-ARRONDISSEMENT
PARIS-2E-ARRONDISSEMENT
PARIS-11E-ARRONDISSEMENT
PARIS-3E-ARRONDISSEMENT
PARIS-6E-ARRONDISSEMENT
PARIS-19E-ARRONDISSEMENT
PARIS-20E-ARRONDISSEMENT
PARIS-15E-ARRONDISSEMENT
PARIS-5E-ARRONDISSEMENT
PARIS-7E-ARRONDISSEMENT
PARIS-8E-ARRONDISSEMENT
PARIS-13E-ARRONDISSEMENT
PARIS-12E-ARRONDISSEMENT
PARIS-18E-ARRONDISSEMENT
PARIS-1ER-ARRONDISSEMENT
PARIS-10E-ARRONDISSEMENT
PARIS-16E-ARRONDISSEMENT
PARIS-4E-ARRONDISSEMENT
PARIS-17E-ARRONDISSEMENT
PARIS-14E-ARRONDISSEMENT


Sample our data

In [70]:
venues_in_Paris.head()


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
1  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
2  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
3  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
4  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   

                    Venue                 Venue Category  
0     Place Saint-Georges                          Plaza  
1                     Qee           Gym / Fitness Center  
2  Le Bouclier de Bacchus                       Wine Bar  
3                  So Nat  Vegetarian / Vegan Restaurant  
4               Aji Dulce          Venezuelan Restaurant

In [71]:
venues_in_Paris.shape


(1329, 5)

1329 records of venues were collected for Paris

### Grouping by Venue Categories

Check how many Venue Catergories there are 

In [72]:
venues_in_Paris.groupby('Venue Category').max()


Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                             
Afghan Restaurant       PARIS-11E-ARRONDISSEMENT               48.859415   
African Restaurant       PARIS-9E-ARRONDISSEMENT               48.886869   
American Restaurant     PARIS-19E-ARRONDISSEMENT               48.892735   
Antique Shop             PARIS-9E-ARRONDISSEMENT               48.876896   
Argentinian Restaurant   PARIS-3E-ARRONDISSEMENT               48.863054   
...                                          ...                     ...   
Wine Bar                 PARIS-9E-ARRONDISSEMENT               48.892735   
Wine Shop                PARIS-3E-ARRONDISSEMENT               48.876029   
Women's Store            PARIS-6E-ARRONDISSEMENT               48.867903   
Zoo                     PARIS-12E-ARRONDISSEMENT               48.835156   
Zoo Exhibit             PARIS-12E-ARRONDISSEMENT               48.835156   

                        Neighbourhood Longitude  \
Venue Category                                    
Afghan Restaurant                      2.378741   
African Restaurant                     2.400820   
American Restaurant                    2.384694   
Antique Shop                           2.337460   
Argentinian Restaurant                 2.359361   
...                                         ...   
Wine Bar                               2.400820   
Wine Shop                              2.361113   
Women's Store                          2.344107   
Zoo                                    2.419807   
Zoo Exhibit                            2.419807   

                                                           Venue  
Venue Category                                                    
Afghan Restaurant                                    Afghanistan  
African Restaurant                             Wally Le Saharien  
American Restaurant                                     Harper's  
Antique Shop                             Hôtel des Ventes Drouot  
Argentinian Restaurant                                     Anahi  
...                                                          ...  
Wine Bar                                        Vingt Vins d'Art  
Wine Shop                                         Trois Fois Vin  
Women's Store                               L'Appartement Sézane  
Zoo                                     Parc zoologique de Paris  
Zoo Exhibit             Grande Serre du Parc Zoologique de Paris  

[206 rows x 4 columns]

With 206 records, Paris is very diverse

### One Hot Encoding

In [73]:
Paris_venue_cat = pd.get_dummies(venues_in_Paris[['Venue Category']], prefix="", prefix_sep="")
Paris_venue_cat

Afghan Restaurant  African Restaurant  American Restaurant  \
0                     0                   0                    0   
1                     0                   0                    0   
2                     0                   0                    0   
3                     0                   0                    0   
4                     0                   0                    0   
...                 ...                 ...                  ...   
1324                  0                   0                    0   
1325                  0                   0                    0   
1326                  0                   0                    0   
1327                  0                   0                    0   
1328                  0                   0                    0   

      Antique Shop  Argentinian Restaurant  Art Gallery  Art Museum  \
0                0                       0            0           0   
1                0                       0            0           0   
2                0                       0            0           0   
3                0                       0            0           0   
4                0                       0            0           0   
...            ...                     ...          ...         ...   
1324             0                       0            0           0   
1325             0                       0            0           0   
1326             0                       0            0           0   
1327             0                       0            0           0   
1328             0                       0            0           0   

      Arts & Crafts Store  Asian Restaurant  Athletics & Sports  ...  \
0                       0                 0                   0  ...   
1                       0                 0                   0  ...   
2                       0                 0                   0  ...   
3                       0                 0                   0  ...   
4                       0                 0                   0  ...   
...                   ...               ...                 ...  ...   
1324                    0                 0                   0  ...   
1325                    0                 0                   0  ...   
1326                    0                 0                   0  ...   
1327                    0                 0                   0  ...   
1328                    0                 0                   0  ...   

      Turkish Restaurant  Udon Restaurant  Vegetarian / Vegan Restaurant  \
0                      0                0                              0   
1                      0                0                              0   
2                      0                0                              0   
3                      0                0                              1   
4                      0                0                              0   
...                  ...              ...                            ...   
1324                   0                0                              0   
1325                   0                0                              0   
1326                   0                0                              0   
1327                   0                0                              0   
1328                   0                0                              0   

      Venezuelan Restaurant  Vietnamese Restaurant  Wine Bar  Wine Shop  \
0                         0                      0         0          0   
1                         0                      0         0          0   
2                         0                      0         1          0   
3                         0                      0         0          0   
4                         1                      0         0          0   
...                     ...                    ...       ...        ...   
1324                      0      

Add Neighborhoods

In [74]:
Paris_venue_cat['Neighbourhood'] = venues_in_Paris['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [Paris_venue_cat.columns[-1]] + list(Paris_venue_cat.columns[:-1])
Paris_venue_cat = Paris_venue_cat[fixed_columns]

Paris_venue_cat.head()

Neighbourhood  Afghan Restaurant  African Restaurant  \
0  PARIS-9E-ARRONDISSEMENT                  0                   0   
1  PARIS-9E-ARRONDISSEMENT                  0                   0   
2  PARIS-9E-ARRONDISSEMENT                  0                   0   
3  PARIS-9E-ARRONDISSEMENT                  0                   0   
4  PARIS-9E-ARRONDISSEMENT                  0                   0   

   American Restaurant  Antique Shop  Argentinian Restaurant  Art Gallery  \
0                    0             0                       0            0   
1                    0             0                       0            0   
2                    0             0                       0            0   
3                    0             0                       0            0   
4                    0             0                       0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  ...  Turkish Restaurant  \
0           0                    0                 0  ...                   0   
1           0                    0                 0  ...                   0   
2           0                    0                 0  ...                   0   
3           0                    0                 0  ...                   0   
4           0                    0                 0  ...                   0   

   Udon Restaurant  Vegetarian / Vegan Restaurant  Venezuelan Restaurant  \
0                0                              0                      0   
1                0                              0                      0   
2                0                              0                      0   
3                0                              1                      0   
4                0                              0                      1   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Zoo  Zoo Exhibit  
0                      0         0          0              0    0            0  
1                      0         0          0              0    0            0  
2                      0         1          0              0    0            0  
3                      0         0          0              0    0            0  
4                      0         0          0              0    0            0  

[5 rows x 207 columns]

### Venue categories mean value

In [75]:
Paris_grouped = Paris_venue_cat.groupby('Neighbourhood').mean().reset_index()
Paris_grouped.head()

Neighbourhood  Afghan Restaurant  African Restaurant  \
0  PARIS-10E-ARRONDISSEMENT            0.00000                0.02   
1  PARIS-11E-ARRONDISSEMENT            0.02381                0.00   
2  PARIS-12E-ARRONDISSEMENT            0.00000                0.00   
3  PARIS-13E-ARRONDISSEMENT            0.00000                0.00   
4  PARIS-14E-ARRONDISSEMENT            0.00000                0.00   

   American Restaurant  Antique Shop  Argentinian Restaurant  Art Gallery  \
0                  0.0           0.0                     0.0          0.0   
1                  0.0           0.0                     0.0          0.0   
2                  0.0           0.0                     0.0          0.0   
3                  0.0           0.0                     0.0          0.0   
4                  0.0           0.0                     0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  ...  Turkish Restaurant  \
0     0.00000                  0.0          0.030000  ...                 0.0   
1     0.02381                  0.0          0.023810  ...                 0.0   
2     0.00000                  0.0          0.000000  ...                 0.0   
3     0.00000                  0.0          0.183333  ...                 0.0   
4     0.00000                  0.0          0.000000  ...                 0.0   

   Udon Restaurant  Vegetarian / Vegan Restaurant  Venezuelan Restaurant  \
0              0.0                       0.010000                    0.0   
1              0.0                       0.047619                    0.0   
2              0.0                       0.000000                    0.0   
3              0.0                       0.000000                    0.0   
4              0.0                       0.000000                    0.0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store       Zoo  \
0                0.00000   0.02000       0.02            0.0  0.000000   
1                0.02381   0.02381       0.00            0.0  0.000000   
2                0.00000   0.00000       0.00            0.0  0.142857   
3                0.20000   0.00000       0.00            0.0  0.000000   
4                0.00000   0.00000       0.00            0.0  0.000000   

   Zoo Exhibit  
0     0.000000  
1     0.000000  
2     0.142857  
3     0.000000  
4     0.000000  

[5 rows x 207 columns]

### Top Venue Categories

In [76]:
# create a new dataframe for Paris
neighborhoods_venues_sorted_paris = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_paris['Neighbourhood'] = Paris_grouped['Neighbourhood']

for ind in np.arange(Paris_grouped.shape[0]):
    neighborhoods_venues_sorted_paris.iloc[ind, 1:] = return_most_common_venues(Paris_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_paris.head()

Neighbourhood  1st Most Common Venue 2nd Most Common Venue  \
0  PARIS-10E-ARRONDISSEMENT      French Restaurant                 Hotel   
1  PARIS-11E-ARRONDISSEMENT             Restaurant                  Café   
2  PARIS-12E-ARRONDISSEMENT            Zoo Exhibit           Supermarket   
3  PARIS-13E-ARRONDISSEMENT  Vietnamese Restaurant      Asian Restaurant   
4  PARIS-14E-ARRONDISSEMENT      French Restaurant                 Hotel   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop                  Café                Bistro   
1     French Restaurant    Italian Restaurant                Bakery   
2                Bistro                  Park   Monument / Landmark   
3    Chinese Restaurant       Thai Restaurant     French Restaurant   
4   Japanese Restaurant     Food & Drink Shop                Bistro   

  6th Most Common Venue          7th Most Common Venue 8th Most Common Venue  \
0    Italian Restaurant              Indian Restaurant           Pizza Place   
1                Bistro  Vegetarian / Vegan Restaurant          Cocktail Bar   
2                   Zoo                           Pool         Deli / Bodega   
3             Juice Bar             Italian Restaurant                 Hotel   
4                  Café             Italian Restaurant                Bakery   

  9th Most Common Venue    10th Most Common Venue  
0      Asian Restaurant            Breakfast Spot  
1           Pastry Shop  Mediterranean Restaurant  
2              Fountain                Food Court  
3              Creperie              Dessert Shop  
4             Brasserie      Fast Food Restaurant

### Model Building

Use K Means Clustering to cluster the city of Paris to 5 

In [77]:
k_num_clusters = 5

Paris_grouped_clustering = Paris_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_Paris = KMeans(n_clusters=k_num_clusters, random_state=0).fit(Paris_grouped_clustering)
kmeans_Paris

KMeans(n_clusters=5, random_state=0)

### Labeling Clustered Data

In [78]:
kmeans_Paris.labels_


array([2, 2, 3, 4, 1, 2, 0, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 2],
      dtype=int32)

In [79]:
neighborhoods_venues_sorted_paris.insert(0, 'Cluster Labels', kmeans_Paris.labels_ +1)

In [80]:
paris_data = paris_combined_data

paris_data = paris_data.join(neighborhoods_venues_sorted_paris.set_index('Neighbourhood'), on='nom_comm')

paris_data.head()

postal_code                  nom_comm nom_dept   Latitude  Longitude  \
0       75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460   
1       75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107   
2       75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741   
3       75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361   
4       75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               3     French Restaurant                 Hotel   
1               3     French Restaurant          Cocktail Bar   
2               3            Restaurant                  Café   
3               3           Coffee Shop     French Restaurant   
4               3     French Restaurant        Chocolate Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                Bakery                Bistro   Japanese Restaurant   
1              Wine Bar    Italian Restaurant           Salad Place   
2     French Restaurant    Italian Restaurant                Bakery   
3                Bakery   Japanese Restaurant          Burger Joint   
4                Bistro                Bakery    Italian Restaurant   

  6th Most Common Venue          7th Most Common Venue 8th Most Common Venue  \
0                Lounge                       Wine Bar                   Bar   
1       Thai Restaurant                         Bakery                   Bar   
2                Bistro  Vegetarian / Vegan Restaurant          Cocktail Bar   
3                Bistro                    Art Gallery          Gourmet Shop   
4    Athletics & Sports                          Hotel              Fountain   

  9th Most Common Venue    10th Most Common Venue  
0          Cocktail Bar               Pizza Place  
1                 Hotel        Mexican Restaurant  
2           Pastry Shop  Mediterranean Restaurant  
3          Cocktail Bar            Sandwich Place  
4           Pastry Shop                       Pub

Drop all NaN values to prevent skew

In [81]:
paris_data_nonan = paris_data.dropna(subset=['Cluster Labels'])


### Visualize the clustered neighborhood

In [82]:
map_clusters_paris = folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(paris_data_nonan['Latitude'], paris_data_nonan['Longitude'], paris_data_nonan['nom_comm'], paris_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + ' ' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.8
        ).add_to(map_clusters_paris)
        
map_clusters_paris

### Examining our Cluster

Cluster 1

In [83]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 1, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]


nom_comm  Cluster Labels 1st Most Common Venue  \
16  PARIS-16E-ARRONDISSEMENT               1                  Lake   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
16                 Plaza                  Pool         Boat or Ferry   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
16           Bus Station              Bus Stop     French Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
16                  Park            Art Museum                  Diner

Cluster 2

In [84]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 2, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]


nom_comm  Cluster Labels 1st Most Common Venue  \
9    PARIS-7E-ARRONDISSEMENT               2     French Restaurant   
10   PARIS-8E-ARRONDISSEMENT               2     French Restaurant   
18  PARIS-17E-ARRONDISSEMENT               2     French Restaurant   
19  PARIS-14E-ARRONDISSEMENT               2     French Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9                  Hotel                  Café    Italian Restaurant   
10                 Hotel           Art Gallery                   Spa   
18                 Hotel                Bakery    Italian Restaurant   
19                 Hotel   Japanese Restaurant     Food & Drink Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9                  Plaza        History Museum           Coffee Shop   
10          Cocktail Bar               Theater           Coffee Shop   
18                  Café   Japanese Restaurant                Bistro   
19                Bistro                  Café    Italian Restaurant   

   8th Most Common Venue       9th Most Common Venue  10th Most Common Venue  
9           Cocktail Bar                  Art Museum            Gourmet Shop  
10         Grocery Store  Modern European Restaurant  Furniture / Home Store  
18                 Plaza                  Restaurant                   Diner  
19                Bakery                   Brasserie    Fast Food Restaurant

Cluster 3

In [85]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 3, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]


nom_comm  Cluster Labels 1st Most Common Venue  \
0    PARIS-9E-ARRONDISSEMENT               3     French Restaurant   
1    PARIS-2E-ARRONDISSEMENT               3     French Restaurant   
2   PARIS-11E-ARRONDISSEMENT               3            Restaurant   
3    PARIS-3E-ARRONDISSEMENT               3           Coffee Shop   
4    PARIS-6E-ARRONDISSEMENT               3     French Restaurant   
5   PARIS-19E-ARRONDISSEMENT               3     French Restaurant   
6   PARIS-20E-ARRONDISSEMENT               3                Bakery   
7   PARIS-15E-ARRONDISSEMENT               3    Italian Restaurant   
8    PARIS-5E-ARRONDISSEMENT               3     French Restaurant   
13  PARIS-18E-ARRONDISSEMENT               3                   Bar   
14  PARIS-1ER-ARRONDISSEMENT               3     French Restaurant   
15  PARIS-10E-ARRONDISSEMENT               3     French Restaurant   
17   PARIS-4E-ARRONDISSEMENT               3     French Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Hotel                Bakery                Bistro   
1           Cocktail Bar              Wine Bar    Italian Restaurant   
2                   Café     French Restaurant    Italian Restaurant   
3      French Restaurant                Bakery   Japanese Restaurant   
4         Chocolate Shop                Bistro                Bakery   
5                    Bar                 Hotel                Bistro   
6                 Bistro                 Plaza   Japanese Restaurant   
7      French Restaurant                 Hotel           Coffee Shop   
8                  Hotel    Italian Restaurant                Bakery   
13     French Restaurant    Italian Restaurant           Pizza Place   
14   Japanese Restaurant                 Hotel                 Plaza   
15                 Hotel           Coffee Shop                  Café   
17        Ice Cream Shop        Clothing Store                 Hotel   

   5th Most Common Venue 6th Most Common Venue          7th Most Common Venue  \
0    Japanese Restaurant                Lounge                       Wine Bar   
1            Salad Place       Thai Restaurant                         Bakery   
2                 Bakery                Bistro  Vegetarian / Vegan Restaurant   
3           Burger Joint                Bistro                    Art Gallery   
4     Italian Restaurant    Athletics & Sports                          Hotel   
5            Supermarket              Beer Bar                        Brewery   
6                    Bar     French Restaurant              Food & Drink Shop   
7                   Park                Bistro                     Restaurant   
8                  Plaza                  Café                            Bar   
13            Restaurant                  Café                         Bistro   
14           Coffee Shop    Italian Restaurant                           Café   
15                Bistro    Italian Restaurant              Indian Restaurant   
17      Pedestrian Plaza                  Park                    Pastry Shop   

   8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
0                    Bar          Cocktail Bar               Pizza Place  
1                    Bar                 Hotel        Mexican Restaurant  
2           Cocktail Bar           Pastry Shop  Mediterranean Restaurant  
3           Gourmet Shop          Cocktail Bar            Sandwich Place  
4               Fountain           Pastry Shop                       Pub  
5     Seafood Restaurant        Sandwich Place                       Spa  
6                   Park                  Café        Italian Restaurant  
7      Indian Restaurant   Japanese Restaurant       Lebanese Restaurant  
8       Greek Restaurant                   Pub            Science Museum  
13                 Plaza                 Hotel         Convenience Store  
14      Ramen Restaurant                Bakery         Korean Restaura

Cluster 4

In [86]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 4, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]


nom_comm  Cluster Labels 1st Most Common Venue  \
12  PARIS-12E-ARRONDISSEMENT               4           Zoo Exhibit   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12           Supermarket                Bistro                  Park   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12   Monument / Landmark                   Zoo                  Pool   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12         Deli / Bodega              Fountain             Food Court

Cluster 5

In [87]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 5, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]


nom_comm  Cluster Labels  1st Most Common Venue  \
11  PARIS-13E-ARRONDISSEMENT               5  Vietnamese Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11      Asian Restaurant    Chinese Restaurant       Thai Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
11     French Restaurant             Juice Bar    Italian Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11                 Hotel              Creperie           Dessert Shop

### Results and Discussion

London has very multicultural neighborhoods.  The cuusines range from Italian, Turkish to Chinese.  There is also a variety of different types of eating places such as restaurants, bars, coffee shops and juice bars. Further, there are numerous shopping options including fish markets, clothing stores, flea markets, glower shops etc.  for leisure, neighborhoods have parks, gold courses, gyms, zoos, etc.  Main methods of transportation are buses and trains. London would provide an entertainig experience.

Paris is much smaller geographically, but has a wide variety of activities and eateries.  The cusines offered are French, Thai, Cambodian, Asian, etc. There are many restaurants and bars, and a numerous amount of bistros.  There are more modes of public transport in paris, namely buses, bikes boats and ferries  In regards to leisure and sight seeing, there are plazas, parks, historic sites, clothing shops, art galleries and museums.  Paris seems like a very diverse interesting vacation spot. 

### Conclusion

This project was meant to showcase what the cities of Paris and London had to offer to potential migrants and tourists. We explored venues present in each location.  Each of the cities neighborhoods have a wide variety of expereinces.  Yet, they offer different experiences.  Both offer a great getaway with many places to explore and numerous cusines to try.  Culture is evident in both cities.  Our findings will help people narrow down what specific experiences they would like and which city matcher their preferences more. 